### ISSUE TITLE
Sending a message that includes a pipe inside of two angle brackets adds message as wrong bot user.

### ISSUE THREAD

**Issue by kallanreed** [slackhq/hubot-slack#344]
* [x] I've read and understood the [Contributing guidelines](./CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](./CODE_OF_CONDUCT.md).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

#### Description
> calling robot.send with a message like `<foo|bar>` causes the message to be added by user "bot" in Slack instead of the configured bot. Because it's a different user than the configured bot user (let's say "my_bot_name") the messages written by "bot" are heard by the hubot.

I noticed the behavior when the hubot started responding to the help text from `@my_bot_name help` which rendered `<foo|bar>` as part of a command help string.

#### Reproducible in:
    "hubot": "^2.19.0",
    "hubot-diagnostics": "0.0.1",
    "hubot-google-images": "^0.2.6",
    "hubot-google-translate": "^0.2.0",
    "hubot-help": "^0.2.0",
    "hubot-heroku-keepalive": "^1.0.2",
    "hubot-maps": "0.0.2",
    "hubot-pugme": "^0.1.0",
    "hubot-redis-brain": "0.0.3",
    "hubot-rules": "^0.1.1",
    "hubot-scripts": "^2.17.2",
    "hubot-shipit": "^0.2.0",
    "hubot-slack": "^4.0.2",
    "node-whois": "^2.1.3"

#### Steps to reproduce:
1. send `"<foo|bar>"` from within a hubot script

### Expected result:
> "\<foo|bar>" to be added to the channel as "my_bot_name"

### Actual result:
> "\<foo|bar>" is added as a user called "bot" which is not even configured in the team settings.

### Attachments:
![image](https://cloud.githubusercontent.com/assets/3761006/17835740/1b82ef9e-672f-11e6-95c5-72441d3b97d5.png)

In the screenshot above, I have two response scripts. The first is a simple echo that calls toLowerCase() before sending and the second responds to "test". In the example you can see that the "test" script is responding the bugged message added by the "bot" caused by this issue.

**Comment by kallanreed** [comment]
Here are some debug logs, it looks like it might be broken in the slack web API itself:
```
[Sun Aug 21 2016 16:47:52 GMT+0000 (Coordinated Universal Time)] DEBUG Received message: '@whatisbot bad' in channel: pr
otect-bot-funland, from: kylereed
[Sun Aug 21 2016 16:47:52 GMT+0000 (Coordinated Universal Time)] DEBUG Message '@whatisbot bad' matched regex //^\s*[@]?
whatisbot[:,]?\s*(?:bad)//; listener.options = { id: null }
[Sun Aug 21 2016 16:47:52 GMT+0000 (Coordinated Universal Time)] DEBUG Executing listener callback for Message '@whatisb
ot bad'
[Sun Aug 21 2016 16:47:52 GMT+0000 (Coordinated Universal Time)] DEBUG Sending to C237Q2NP7: <bad|robot>
[Sun Aug 21 2016 16:47:52 GMT+0000 (Coordinated Universal Time)] DEBUG Received message: '<bad|robot>' in channel: prote
ct-bot-funland, from: hubot
[Sun Aug 21 2016 16:47:52 GMT+0000 (Coordinated Universal Time)] DEBUG No listeners executed; falling back to catch-all
```

**Comment by kallanreed** [comment]
After looking at formatter.coffee in the hubot-slack package, it looks like Slack treats `<foo|bar>` as an ID and a label which may be part of the reason this is getting messed up. Not sure if there's a way to escape the strings so it doesn't get interpreted by Slack.

**Comment by kallanreed** [comment]
It looks like this can be fixed by escaping brackets and ampersands before sending the message to Slack.

Add the following to the formatter mentions method.

```
  ### 
  Recursively replace @username with <@UXXXXX> for mentioning users and channels
  ###
  mentions: (text) ->
    return if text is null # nothing to do

    text = text.replace /&/g, "&amp;"
    text = text.replace /</g, "&lt;"
    text = text.replace />/g, "&gt;"
```

**Comment by DEGoodmanWilson** [comment]
Yeah, I think I know how to fix this…which is to hand off parsing of links to the Slack backend.

**Comment by kallanreed** [comment]
That sounds right. I attempted a fix in the formatter, but it was going to need special handling of the @user and !here entities.

Sent from my iPhone

On Aug 24, 2016, at 23:24, Don Goodman-Wilson <notifications@github.com<mailto:notifications@github.com>> wrote:


Yeah, I think I know how to fix this...which is to hand off parsing of links to the Slack backend.

-

**Comment by DEGoodmanWilson** [comment]
This should now be fixed as of the latest release, 4.0.3, going out momentarily. Can you verify for me?

**Comment by iancward** [comment]
@DEGoodmanWilson when I ask for help (via the hubot-help script), my bot now has my user ID in angle brackets instead of my user name.

e.g.:
```
<@U123456>: my_botname: a_fun_command
```

**Comment by DEGoodmanWilson** [comment]
:facepalm:

**Comment by DEGoodmanWilson** [comment]
@iancward can you provide a snippet that reproduces the problem? I can't seem to repro it locally. Indeed, this seems a new issue, can you open a new issue, as the original is now fixed?

**Comment by DEGoodmanWilson** [comment]
This should be fixed in 4.0.4

**Comment by iancward** [comment]
@DEGoodmanWilson thanks! it looks good now. 
And thanks for the humorous CHANGELOG entries. :-D
